first_order_model/shrink_util.py

```diff
@@ -0,0 +1,1235 @@
+                y, generator_type)
+            loss_values = [val.mean() for val in losses_generator.values()]
+            loss = sum(loss_values)
+            total_loss += loss.item()
+
+    return total_loss
+
+def get_following_layers_skip_depthwise(following_layers, layer_graph, x):
+    """
+    Specialized version of follo that skips over depthwise convolved layers. Only used when sorting the outputs of a layer for importance.
+
+    """
+    following_layers.append(x.index)
+    if x.type == 'conv' and x.value.groups == 1:
+        return
+    for y in x.after:
+        get_following_layers_skip_depthwise(following_layers, layer_graph, layer_graph[y])
+
+def follow(following_layers, layer_graph, x):
+    """
+    Get any possible layers which are directly impacted by editing x's outputs
+    """
+    following_layers.append(x.index)
+    if x.type == 'conv':
+        return
+    for y in x.after:
+        follow(following_layers, layer_graph, layer_graph[y])
+
+
+def get_first_conv(layer_graph, following_layers):
+    """
+    Gets first convolution in following layers
+    """
+    for layer in following_layers:
+        if layer_graph[layer].type == 'conv' and layer_graph[layer].value.groups == 1:
+            return layer
+
+def compute_deletion(layer_graph,
+                     custom_deletions,
+                     deleted_things,
+                     layer,
+                     sort,
+                     custom,
+                     reason=None):
+    """
+    Given a layer, generate the list of the indexes we need to delete from its following layers
+    """
+    # find all the following layers
+    following_layers = []
+
+    for after_layer in layer_graph[layer].after:
+        follow(following_layers, layer_graph, layer_graph[after_layer])
+
+    # If you delete a part of the previous one, not just all of it then you cannot just say these n are to be deleted
+    # So for each layer you need to figure out what you are deleting from it, then for the next layer figure out what is deleted
+
+    deletions = {}
+    if isinstance(custom, list):
+        deletions[layer] = custom
+    elif isinstance(custom, int):
+        # The sorting magic happens here
+        # Find the first convolution
+        if sort:
+            if layer_graph[layer].type == 'bn':
+                breakpoint()
+            depthwise_skipped_following_layers = []
+            for after_layer in layer_graph[layer].after:
+                get_following_layers_skip_depthwise(depthwise_skipped_following_layers, layer_graph, layer_graph[after_layer])
+
+            first_conv = get_first_conv(layer_graph, depthwise_skipped_following_layers)
+
+            if first_conv is None:
+                breakpoint()
+
+            slices = get_relevent_slice(layer_graph, first_conv, layer)
+            if len(slices) == 0:
+                breakpoint()
+
+            importances = torch.zeros(layer_graph[layer].o).cuda()
+            for single_slice in slices:
```

Comment:
not sure exactly what each slice represents, why the importances are being added

Reply:
Second case of resolving a comment without explaining. Please don't make me do my job twice and try to piece together what you've resolved, and what you haven't

Reply:
Should be explained in the code now.